### PR TITLE
Tuple optional

### PR DESCRIPTION
**Pull request by derekdreery**
This patch adds 2 features. Special care is needed with the `Optional` widget. While I'm sure it is very useful, I'm not sure I've implemented it correctly.

### CODE REVIEW

**Review by Maan2003 — COMMENTED** [review]
Tuple lens looks good! For `Optional`, I recommend moving [Runebender's `Maybe`](https://github.com/linebender/runebender/blob/f38c736/runebender-lib/src/widgets/maybe.rs) to druid

**Comment by cmyr** [comment]
I have similar doubts about the correctness of my `Maybe` widget, but I've worked out at least some of the bugs, and it is certainly useful.

**Comment by derekdreery** [comment]
@cmyr could you comment on the difference between this `Optional` and runebender's `Maybe`? I didn't include a `None` option here, since you can get that functionality using `Either`, but maybe it makes more sense to include it. Also, how to handle events where the event should go to the widget (`should_go_to_hidden`, or something like that, and lifecycle events), but we don't have any data to send to the widget? I handled it by just sending the default value of the `T` of `Option<T>` to the some widget when the `data` was `None`, but this seems suboptimal.

**Comment by derekdreery** [comment]
And another question, if we have a widget for the `None` case, could we include a `Null` widget that lays itself out as `bc.min()` and does nothing else? It would go nicely with `Maybe`/`Optional`.

**Comment by cmyr** [comment]
in no order:

- there sort of is a `Null` widget, in the form of `SizedBox::empty`.
- I like letting there be a `None` option, in case you want a custom placeholder.
- I don't have any good answer about `should_go_to_hidden`. I think the solution I would advocate to not have the widget be hidden, but instead to create and destroy it as needed.

**Comment by derekdreery** [comment]
About the `SizedBox::empty` - it looks to be equivalent to a theoretical `Null` widget. My question is: how do we make this use of `SizedBox` discoverable?

**Comment by cmyr** [comment]
I'm open to suggestions!

**Comment by derekdreery** [comment]
@cmyr I've swapped out the `Optional` for runebender's `Maybe`, and added some intra-crate doc links.

**Comment by derekdreery** [comment]
Ready for review :)

File: druid/src/widget/sized_box.rs [hunk 1]
```diff
@@ -45,6 +45,10 @@ impl<T> SizedBox<T> {
     }
 
     /// Construct container without child, and both width and height not set.
+    ///
+    /// If the widget is unchanged, it will do nothing, which can be useful if you want to draw a
+    /// widget some of the time (for example, it is used to implement
+    /// [`Maybe`][crate::widget::Maybe]).
     pub fn empty() -> Self {
```
Comment (Maan2003) [inline druid/src/widget/sized_box.rs:52]: adding `#[doc(alias = "null")]` will help discoverability

Comment (derekdreery) [reply druid/src/widget/sized_box.rs:52]: Ack I was genuinely going to do that but forgot. Thanks!

File: druid/src/lens/lens.rs [hunk 2]
```diff
@@ -567,3 +567,31 @@ impl<A, B: Clone> Lens<A, B> for Constant<B> {
         f(&mut tmp)
     }
 }
+
+/// A lens that combines two lenses into a tuple.
+#[derive(Debug, Copy, Clone)]
+pub struct Tuple<L1, L2>(pub L1, pub L2);
```
Comment (rjwittams) [inline druid/src/lens/lens.rs:573]: Wonder if this should be Tuple2 or Pair, as it doesn't apply to all tuples

**Comment by cmyr** [comment]
Do we need both `Maybe` *and* `Optional`? In my use, `Maybe` covers the optional category well enough.

**Comment by derekdreery** [comment]
> 
> 
> Do we need both `Maybe` _and_ `Optional`? In my use, `Maybe` covers the optional category well enough.

Oh I thought I had removed `Optional`. Will do.

**Comment by derekdreery** [comment]
Ready for review again.

**Review by cmyr — APPROVED** [review]
I won't make strong guarantees about covering all the corner cases in `Maybe`, but I haven't seen a crash in runebender in a while? 😬

**Pull request merged**